feat: Create a backend to transform CSV files to DoclingDocument

tobiasstrebitzer opened the pull request:
**Description of the changes:**
- Add CsvDocumentBackend to (1) detect the dialect of a csv file, (2) parse rows from CSV files and to (3) convert them to a DoclingDocument.
- Add tests for CSV conversion of various dialects (./tests/test_backend_csv.py).
- Add tests for CSV format detections (./tests/test_input_doc.py).

**Known limitations:**
- Currently, the representation of CSV as Markdown (default output) is not a great strategy for generating embedding vectors or as context for GenAI, so this is only of limited use for AI / RAG purposes. This could be overcome by creating a specialized export format for tabular data tailored for GenAI / RAG use cases.

**Issue resolved by this Pull Request:**
Resolves #646

**Checklist:**

- [x] Documentation has been updated, if necessary.
- [x] Examples have been added, if necessary.
- [x] Tests have been added, if necessary.
- [x] Code style and formatting applied via pre-commit hook.

dolfim-ibm commented:
@tobiasstrebitzer I think you still have to apply the pre-commit hooks which will style the code automatically. For example via `poetry run pre-commit run --all-files`.

The PR implements an interesting feature. I was wondering how flexible is it regarding the CSV flavours? e.g. `,`, `;`, or other separators?

tobiasstrebitzer commented:
@dolfim-ibm thanks for your feedback, and great catch regarding various CSV flavours.

I have updated my PR:
- fixed code-style via pre-commit
- implemented automatic parsing of csv dialect
- added additional tests for various csv dialects (comma, semicolon, pipe, tab)

PeterStaar-IBM commented:
@tobiasstrebitzer Very nice addition!

two minor comments: 

1. could you test that we do not run into trouble when we have comma's int the values. I think that typically the values are then quoted. It is a common error that occurs. Maybe we could add a test for that. If this is not supported, I dont mind, as long as we raise an exception (so we know when something went wrong).
2. If the csv is invalid (inconsistent number of rows), I suppose that the `csv` package will throw an error?

super stoked to get this merged in soon!

tobiasstrebitzer commented:
@PeterStaar-IBM thanks for your comments,

---

> **1. could you test that we do not run into trouble when we have comma's int the values. I think that typically the values are then quoted. It is a common error that occurs. Maybe we could add a test for that. If this is not supported, I dont mind, as long as we raise an exception (so we know when something went wrong).**

I've had a look and created additional tests. The `csv` package takes care of this pretty well.

There can still be user errors, e.g.:
```csv
d , c , s
. , , , ;
```

In this case, the parser will interpret this as:
```python
[
  ["d", "c", "s"],   # length = 3
  [".", "", "", ";"] # length = 4
]
```

But at least, a warning will be raised due to inconsistent column lengths:

_Inconsistent column lengths detected in CSV data. Expected 3 columns, but found rows with varying lengths. Ensure all rows have the same number of columns._

---

> **2. If the csv is invalid (inconsistent number of rows), I suppose that the csv package will throw an error?**

It seems the `csv` package allows inconsistent column lengths (even in `strict` mode). I'm raising a warning instead (test included).

---

**I have updated my PR:**
- Improved dialect detection, avoiding expensive memory allocation
- Additional tests for csv warnings (e.g. inconsistent column lengths)

PeterStaar-IBM commented:
> @PeterStaar-IBM thanks for your comments,
> 
> > **1. could you test that we do not run into trouble when we have comma's int the values. I think that typically the values are then quoted. It is a common error that occurs. Maybe we could add a test for that. If this is not supported, I dont mind, as long as we raise an exception (so we know when something went wrong).**
> 
> I've had a look and created additional tests. The `csv` package takes care of this pretty well.
> 
> There can still be user errors, e.g.:
> 
> ```
> d , c , s
> . , , , ;
> ```
> 
> In this case, the parser will interpret this as:
> 
> ```python
> [
>   ["d", "c", "s"],   # length = 3
>   [".", "", "", ";"] # length = 4
> ]
> ```
> 
> But at least, a warning will be raised due to inconsistent column lengths:
> 
> _Inconsistent column lengths detected in CSV data. Expected 3 columns, but found rows with varying lengths. Ensure all rows have the same number of columns._
> 
> > **2. If the csv is invalid (inconsistent number of rows), I suppose that the csv package will throw an error?**
> 
> It seems the `csv` package allows inconsistent column lengths (even in `strict` mode). I'm raising a warning instead (test included).
> 
> **I have updated my PR:**
> 
> * Improved dialect detection, avoiding expensive memory allocation
> * Additional tests for csv warnings (e.g. inconsistent column lengths)

OK, this looks very good! Let us run the tests and get it going!

PeterStaar-IBM reviewed (approved):
Beautiful work! Thanks for this effort!
